### PR TITLE
fix msms sweep test in verification.py

### PR DESCRIPTION
**Pull request by gbinux**
To address #1365 

For neutral conditions, `msms` and `msprime` correspond nicely.
For selective sweeps,  `msms` and `discoal` are pretty consistent. There are some discrepancies between `msms` and `msprime`, and between `discoal` and `msprime`, but I think it is reasonable as the selective sweep simulation is implemented in a new framework.  Please help take a look @andrewkern .

[MsmsSweeps.zip](https://github.com/tskit-dev/msprime/files/6174514/MsmsSweeps.zip)

### CODE REVIEW

File: verification.py [hunk 1]
```diff
@@ -1269,30 +1263,39 @@ def _msms_params_to_run_msp(self, params):
...
-            sample_size=params["nsam"],
-            Ne=0.25,
-            length=params["num_sites"],
-            mutation_rate=mutation_rate,
-            recombination_rate=recombination_rate,
+        repeats = msprime.sim_ancestry(
+            samples=params["nsam"],
+            population_size=params["refsize"],
             model=model,
+            recombination_rate=recombination_rate,
```
Comment (jeromekelleher) [inline verification.py:1285]: For a comparison with msms I think we should probably use a continuous genome - set ``discrete_genome=False`` here and in the ``sim_mutations`` below.

**Review by jeromekelleher — COMMENTED** [review]
Thanks @gbinux - let's try the continuous genome and see if the discrepancies remain.

File: verification.py (hunk 1 above)
Comment (gbinux) [reply verification.py:1285]: just added this option and the result is similar  @jeromekelleher

File: verification.py [hunk 2]
```diff
@@ -1269,30 +1263,40 @@ def _msms_params_to_run_msp(self, params):
...
+            discrete_genome=False,
+            sequence_length=params["num_sites"],
+            num_replicates=params["nrep"],
         )
 
-        return ts
+        # Critical to use BinaryMutationModel and get ancestral and derived alleles
+        mutated_repeats = [
+            msprime.sim_mutations(
+                ts, rate=mutation_rate, model=msprime.BinaryMutationModel()
```
Comment (jeromekelleher) [inline verification.py:1294]: Need discrete_genome=False here too @gbinux

**Review by jeromekelleher — COMMENTED** [review]
STill need discrete_genome for mutations @gbinux - because the sequence is so short this will make a difference.

File: verification.py (hunk 2 above)
Comment (gbinux) [reply verification.py]: Oops, missed the one in sim_mutations. Just added it as well. @jeromekelleher 

[MsmsSweeps_2.zip](https://github.com/tskit-dev/msprime/files/6175848/MsmsSweeps_2.zip)

File: verification.py [hunk 3]
```diff
@@ -1269,30 +1263,43 @@ def _msms_params_to_run_msp(self, params):
...
 
-        ts = msprime.simulate(
-            sample_size=params["nsam"],
-            Ne=0.25,
-            length=params["num_sites"],
-            mutation_rate=mutation_rate,
-            recombination_rate=recombination_rate,
+        repeats = msprime.sim_ancestry(
+            samples=params["nsam"],
+            population_size=params["refsize"],
```
Comment (jeromekelleher) [inline verification.py:1283]: Ah - these are diploid sample **individuals** now, we need ``params["nsam"] / 2`` here.

If this number is odd, then you can do
```
samples=msprime.SampleSet(params["nsam"], ploidy=1)
```
to achieve the same effect (i.e., keeping the timescale diploid but sampling haploid individuals)

**Review by jeromekelleher — COMMENTED** [review]
These are major differences in my view @gbinux - but I think the samples thing might fix it.

File: verification.py (hunk 3 above)
Comment (andrewkern) [reply verification.py:1283]: yeah this will probably account for a lot of the differences in these tests @gbinux

Comment (gbinux) [reply verification.py:1283]: Good catch! I tried `params["nsam"] / 2`  and avoided using odd numbers. It seems the discrepancies are still there. 
[MsmsSweeps_3.zip](https://github.com/tskit-dev/msprime/files/6190809/MsmsSweeps_3.zip)

Before I keep pushing the code, I want to show my understanding of the relevant parameters/options. Maybe listing them out could help find the issue if it is obvious. The following linked docs is also a draft related to   #1286 

[Compare args/options between different coalescent simulators](https://docs.google.com/document/d/1TmkfWN5hREOj0wy_B6K3OALAYIydgdPTaulImKheQ18/edit?usp=sharing)

Are there additional misunderstandings about the parameters? Is this kind of information useful to be included in the docs? If the discrepancies persist after we excluded the parameter scaling issue, are we going to keep this Test class within the msprime 1.0 release? @jeromekelleher @andrewkern

Comment (jeromekelleher) [reply verification.py:1283]: Hmm, these still look like pretty big differences to me. Your scalings look about right to me in the doc, but I'm not that well up on these other programs any more.

Can you summarise what you think the differences are, and where we're seeing deviations between the programs please?

Comment (andrewkern) [reply verification.py:1283]: Hi @gbinux-- we already know that discoal, msprime, and slim give the same results from my additions to `verification.py`. you might just look in the `DiscoalSweep` and `SweepVsSlim` test classes in there to see where your parameterization conversion differs.

File: verification.py [hunk 4]
```diff
@@ -1269,30 +1263,43 @@ def _msms_params_to_run_msp(self, params):
                     position=params["sel_pos"],
```
Comment (gbinux) [inline verification.py:1263]: Just realized this is another source of differences: for selection position, msms/discaol use a value between 0-1, msprime use position in bp.
After correcting this, it seems much better. This is `test_selective_discoal_vs_msp_use_discoal_paper_param` result:

before:
image

and  after scaling the position

image

**Comment by gbinux** [comment]
I think the new results are good. Here are representative comparisons.
1. `test_selective_discoal_vs_msp_use_discoal_paper_param`
image

2.  `test_selective_msms_vs_discoal_use_discoal_paper_param`
image

3. `test_selective_msms_vs_msp_use_discoal_paper_param`
image
 
@jeromekelleher @andrewkern

**Comment by andrewkern** [comment]
looks great to me @gbinux!

**Review by andrewkern — APPROVED** [review]

**Comment by jeromekelleher** [comment]
These look much better thanks @gbinux! I'm still a little bit worried about the wobble we're seeing on D and pi, but maybe that's something to do with mutation models.

Can you bring this PR up to date with a rebase please @gbinux? I'll merge then.

**Comment by gbinux** [comment]
> These look much better thanks @gbinux! I'm still a little bit worried about the wobble we're seeing on D and pi, but maybe that's something to do with mutation models.
> 
> Can you bring this PR up to date with a rebase please @gbinux? I'll merge then.

Yep. I used the new model syntax and rebased it already before I did the push, so it should be good to go @jeromekelleher 

I observed that the D and Pi differences occur mainly in `msms` vs `discoal` and `msms` vs `msp` but not obviously in `discoal` vs `msp` and so I think the difference due to the internal logics of how `msms` vs `discoal`(`msp`) simulate this type of selection.

**Comment by andrewkern** [comment]
yes it looks like `msms` is the outlier here.

**Review by jeromekelleher — APPROVED** [review]

**Pull request merged**